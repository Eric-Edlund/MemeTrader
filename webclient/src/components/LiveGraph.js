import React, { useEffect, useState } from "react";
import { Skeleton, ToggleButton, ToggleButtonGroup } from "@mui/material";
import { Line } from "react-chartjs-2";
import {
  Chart as ChartJS,
  TimeScale,
  LinearScale,
  PointElement,
  LineElement,
  Tooltip,
} from "chart.js";
import { useTranslation } from "react-i18next";
import { getPriceHistory } from "./api";
import { globalState } from "../App";

ChartJS.register(TimeScale, LinearScale, PointElement, LineElement, Tooltip);

const LiveGraph = ({ stockId, reloadTrigger }) => {
  const { connected } = globalState;
  const { t } = useTranslation();

  const [stockData, setStockData] = useState(null);
  const [range, setRange] = useState(1); // Days
  const [chartOptions, setChartOptions] = useState({});
  function getStartDate() {
    return new Date(new Date().setDate(new Date().getDate() - range));
  }
  const endDate = new Date();

  useEffect(() => {
    setChartOptions({
      aspectRatio: 16 / 9,
      animation: false,
      scales: {
        x: {
          display: true,
          time: {
            timezone: "UTC",
          },

          drawOnChartArea: false,
          drawTicks: true,
          type: "time",
          min: getStartDate(),
          max: new Date(),
          grid: {
            color: () => "rgba(0,0,0,0)",
          },
        },
        y: {
          beginAtZero: true,
          suggestedMax: 100,
          ticks: {
            callback: function (val) {
              return (val / 100).toLocaleString("en-US", {
                style: "currency",
                currency: "USD",
              });
            },
          },
        },
      },
    });
  }, [range]);

  async function getStockData(/** @type {number}*/ stockId) {
    const data = (
      await getPriceHistory(stockId, getStartDate(), endDate)
    ).points.map(({ time, price }) => {
      return { x: new Date(time), y: price };
    });

    // Extrapolate to now
    data.push({ x: new Date(), y: data[data.length - 1].y });

    setStockData({
      datasets: [
        {
          data: data,
          borderColor: "red",
          borderWidth: 2,
          pointRadius: 0,
        },
      ],
    });
  }

  const click = (event) => event.target.click();

  useEffect(() => {
    getStockData(stockId);
  }, [stockId, range, reloadTrigger]);

  return (
    <>
      <div>
        {stockData ? (
          <Line data={stockData} options={chartOptions} />
        ) : (
          <Skeleton variant="rectangular" width="100%" height="30em" />
        )}
      </div>

      <ToggleButtonGroup
        exclusive
        value={"" + range}
        fullWidth
        onChange={function (event) {
          setRange(event.target.value);
        }}
      >
        <ToggleButton value="1" onMouseDown={click}>
          {t("1D")}
        </ToggleButton>
        <ToggleButton onMouseDown={click} value="7">{t("1W")}</ToggleButton>
        <ToggleButton onMouseDown={click} value="30">{t("1M")}</ToggleButton>
        <ToggleButton onMouseDown={click} value="90">{t("3M")}</ToggleButton>
        <ToggleButton onMouseDown={click} value="365">{t("1Y")}</ToggleButton>
        <ToggleButton onMouseDown={click} value="1825">{t("5Y")}</ToggleButton>
        <ToggleButton onMouseDown={click} value="99999">{t("MAX")}</ToggleButton>
      </ToggleButtonGroup>
    </>
  );
};
export default LiveGraph;
