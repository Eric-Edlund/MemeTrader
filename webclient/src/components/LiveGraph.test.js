// import React from 'react';
// import { render, waitFor } from '@testing-library/react';
// import LiveGraph from './LiveGraph';
// import fetchMock from 'jest-fetch-mock';
// import { API_URL } from '../constants';
//
// beforeEach(() => {
//   fetchMock.resetMocks();
// });
//
// it('renders the component', async () => {
//   fetchMock.mockResponseOnce(JSON.stringify({ points: [] }));
//   const component = await render(<LiveGraph stockId={1} reloadTrigger={0} />);
//   expect(component).toMatchSnapshot();
// });
//
// it('renders the error message', async () => {
//   fetchMock.mockRejectOnce(new Error('Error'));
//   const component = await render(<LiveGraph stockId={1} reloadTrigger={0} />);
//   await waitFor(() => expect(component.getByText('Error')).toBeInTheDocument());
// });

// TODO: pointless awaits and constant import

import React from 'react';
import { render, waitFor } from '@testing-library/react';
import LiveGraph from './LiveGraph';
import fetchMock from 'jest-fetch-mock';

beforeEach(() => {
  fetchMock.resetMocks();
});

it('renders the component', async () => {
  fetchMock.mockResponseOnce(JSON.stringify({ points: [] }));
  const component = await render(<LiveGraph stockId={1} reloadTrigger={0} />);
  expect(component).toMatchSnapshot();
});

it('renders the error message', async () => {
  fetchMock.mockRejectOnce(new Error('Error'));
  const component = await render(<LiveGraph stockId={1} reloadTrigger={0} />);
  await waitFor(() => expect(component.getByText('Error')).toBeInTheDocument());
});
