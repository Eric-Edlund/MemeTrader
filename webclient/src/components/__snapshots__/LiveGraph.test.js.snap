// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
        />
      </div>
      <div
        class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-1dy45qy-MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-pressed="true"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root Mui-selected MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="1"
        >
          1D
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="7"
        >
          1W
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="30"
        >
          1M
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="90"
        >
          3M
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="365"
        >
          1Y
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="1825"
        >
          5Y
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-pressed="false"
          class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
          tabindex="0"
          type="button"
          value="99999"
        >
          MAX
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <span
        class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
      />
    </div>
    <div
      class="MuiToggleButtonGroup-root MuiToggleButtonGroup-fullWidth css-1dy45qy-MuiToggleButtonGroup-root"
      role="group"
    >
      <button
        aria-pressed="true"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root Mui-selected MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="1"
      >
        1D
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="7"
      >
        1W
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="30"
      >
        1M
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="90"
      >
        3M
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="365"
      >
        1Y
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="1825"
      >
        5Y
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-17ndd9v-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="99999"
      >
        MAX
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
